x-pack/filebeat/input/streaming/crowdstrike.go

```diff
@@ -0,0 +1,288 @@
+		FeedURL string `json:"dataFeedURL"`
+		Session struct {
+			Token   string    `json:"token"`
+			Expires time.Time `json:"expiration"`
+		} `json:"sessionToken"`
+		RefreshURL   string `json:"refreshActiveSessionURL"`
+		RefreshAfter int    `json:"refreshActiveSessionInterval"`
+	}
+	var body struct {
+		Resources []resource     `json:"resources"`
+		Meta      map[string]any `json:"meta"`
+	}
+	err = dec.Decode(&body)
+	if err != nil {
+		return state, Warning{fmt.Errorf("failed to decode discover body: %w", err)}
+	}
+	s.log.Debugw("stream discover metadata", "meta", mapstr.M(body.Meta))
+
+	var offset int
+	if cursor, ok := state["cursor"].(map[string]any); ok {
+		switch off := cursor["offset"].(type) {
+		case int:
+			offset = off
+		case float64:
+			offset = int(off)
+		}
+	}
+
+	for _, r := range body.Resources {
+		refreshAfter := time.Duration(r.RefreshAfter) * time.Second
+		go func() {
+			const grace = 5 * time.Minute
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-time.After(refreshAfter - grace):
+					req, err := http.NewRequestWithContext(ctx, http.MethodPost, r.RefreshURL, nil)
+					if err != nil {
+						s.metrics.errorsTotal.Inc()
+						s.log.Errorw("failed to prepare refresh stream request", "error", err)
+						return
+					}
+					req.Header.Set("Content-Type", "application/json")
+					resp, err := cli.Do(req)
+					if err != nil {
+						s.metrics.errorsTotal.Inc()
+						s.log.Errorw("failed to refresh stream connection", "error", err)
+						return
+					}
+					err = resp.Body.Close()
+					if err != nil {
+						s.metrics.errorsTotal.Inc()
+						s.log.Warnw("failed to close refresh response body", "error", err)
+					}
+				}
+			}
+		}()
+
+		if offset > 0 {
+			feedURL, err := url.Parse(r.FeedURL)
+			if err != nil {
+				log.Fatalf("failed to parse feed url: %v", err)
+			}
+			feedQuery, err := url.ParseQuery(feedURL.RawQuery)
+			if err != nil {
+				log.Fatalf("failed to parse feed query: %v", err)
+			}
+			feedQuery.Set("offset", strconv.Itoa(offset))
+			feedURL.RawQuery = feedQuery.Encode()
+			r.FeedURL = feedURL.String()
+		}
+
+		req, err := http.NewRequestWithContext(ctx, "GET", r.FeedURL, nil)
+		if err != nil {
+			return state, Warning{fmt.Errorf("failed to make firehose request to %s: %w", r.FeedURL, err)}
+		}
+		req.Header = make(http.Header)
+		req.Header.Add("Accept", "application/json")
+		req.Header.Add("Authorization", "Token "+r.Session.Token)
+
+		resp, err := s.plainClient.Do(req)
+		if err != nil {
+			return state, Warning{fmt.Errorf("failed to get firehose from %s: %w", r.FeedURL, err)}
+		}
+		defer resp.Body.Close()
+
+		dec := json.NewDecoder(resp.Body)
+		for {
+			var msg json.RawMessage
+			err := dec.Decode(&msg)
+			if err != nil {
+				s.metrics.errorsTotal.Inc()
+				if err == io.EOF {
+					s.log.Info("stream ended, restarting")
+					return state, nil
+				}
+				return state, Warning{fmt.Errorf("error decoding event: %w", err)}
+			}
```

Comment:
An EOF just signals the end of the stream, looking at how it's handled, I believe it should not be counted as an error for metrics.

```suggestion
			if err != nil {
				if err == io.EOF {
					s.log.Info("stream ended, restarting")
					return state, nil
				}
				s.metrics.errorsTotal.Inc()
				return state, Warning{fmt.Errorf("error decoding event: %w", err)}
			}
```